src/bsoncxx/builder/list.hpp

```diff
@@ -29,148 +30,119 @@ namespace builder {
 using namespace bsoncxx::types;
 
 ///
-/// A JSON-like builder for creating documents and arrays.
+/// A JSON-like builder for creating BSON documents.
 ///
-class list {
-    using initializer_list_t = std::initializer_list<list>;
+class document {
+    using key_type = stdx::string_view;
+    using mapped_type = document;
+    using value_type = std::pair<key_type, mapped_type>;
+    using initializer_list_t = std::initializer_list<value_type>;
 
    public:
     ///
     /// Creates an empty document.
     ///
-    list() : list({}) {}
+    document() : document({}){};
 
     ///
-    /// Creates a bsoncxx::builder::list from a value of type T. T must be a
-    /// bsoncxx::types::bson_value::value or implicitly convertible to a
-    /// bsoncxx::types::bson_value::value.
+    /// Creates a bsoncxx::builder::document from a value of type T.
     ///
-    /// @param value
-    ///     the BSON value
+    /// @warning T must be a BSON type, i.e., implicitly convertible to a
+    /// bsoncxx::types::bson_value::value.
     ///
     /// @see bsoncxx::types::bson_value::value.
+    //
+    /// @warning Extracting a BSON document from an object created with this constructor will result
+    /// in an empty document. Use the key-value or std::initializer_list constructors instead.
+    ///
+    /// @note This constructor bridges the gap between a homogeneous container, a
+    /// std::initializer_list, and a heterogeneous one, a BSON document. In other words, the
+    /// std::initializer_list constructor implicitly uses this to convert all elements into a common
+    /// type, a bsoncxx::builder::document.
     ///
     template <typename T>
-    list(T value) : val{value} {}
+    document(T v) : _value{v} {}
```

Comment:
It sounds like users should never be calling this constructor directly.  Can we make it private?

Reply:
If we make this private, we can't use it publicly for implicit conversions. It shouldn't really be called directly. The only use case I can think of is this
```
        list::document num = 1;
        list::document doc{"foo", std::move(num)};
```
That being said, I don't there's a way to make a constructor implicit-only in C++. We'd need an `implicit` keyword modifier that behaves oppositely of `explicit`.

Reply:
Edit: I was able to remove this. Now the only constructors are the templated key-value constructors and the init list constructor.